src/instructlab/model/train.py

```diff
@@ -521,20 +609,320 @@ def get_files(directory: str, pattern: str) -> list[str]:
+        output_dir=eval_cache_path,
+        merge_system_user_message=True,
+    )
+
+    # TODO: Below, gpus=device_count() and max_workers=cpu_count() should be parameterized
+    #   by the user.
+
+    server = None
+    model_serve_url = None
+    try:
+        logger.debug("Starting model server for mt-bench answer generation")
+        server, model_serve_url = launch_server(
+            ctx=ctx,
+            model=ckpt_path,
+            model_name=ckpt_path,
+            gpus=torch.cuda.device_count(),
+            max_workers=multiprocessing.cpu_count(),
+            enable_serving_output=False,
+            backend=backends.VLLM,
+        )
+        logger.debug("Generating mt-bench answers")
+        evaluator.gen_answers(model_serve_url)
+    finally:
+        if server is not None:
+            server.shutdown()
+
+    try:
+        logger.debug("Starting model server for mt-bench answer judgement")
+        server, model_serve_url = launch_server(
+            ctx=ctx,
+            model=judge_name,
+            model_name=judge_name,
+            gpus=torch.cuda.device_count(),
+            max_workers=multiprocessing.cpu_count(),
+            backend=backends.VLLM,
+            enable_serving_output=False,
+        )
+        logger.debug("Judging mt-bench answers")
+        mt_bench_results: tuple = evaluator.judge_answers(model_serve_url)
+        ckpt_score: float = mt_bench_results[0]
+    finally:
+        if server is not None:
+            server.shutdown()
+
+    return ckpt_score
+
+
+def _evaluate_dir_of_checkpoints(
+    checkpoints_dir_path: str, eval_func: typing.Callable[..., float]
+) -> tuple[float, str]:
+    """Run eval_func on all model checkpoints in a directory, returning the best performing."""
+
+    # TODO: parallelize MMLU over available GPUs
+
+    results: list[typing.Tuple[float, str]] = []
+    directory = Path(checkpoints_dir_path)
+    for ckpt_path in directory.iterdir():
+        if ckpt_path.is_dir():
+            logger.debug(ckpt_path)
+            ckpt_score = eval_func(ckpt_path=str(ckpt_path))
             click.secho(
-                f"Error while executing training library {exc}",
+                f"CHECKPOINT EVALUATION: {ckpt_path} SCORED {ckpt_score}",
                 fg="red",
+                bg="cyan",
             )
-            raise click.exceptions.Exit(1)
+            results.append((ckpt_score, str(ckpt_path)))
+
+    if not results:
+        raise RuntimeError(
+            "_evaluate_dir_of_checkpoints - No checkpoints were evaluated successfully at %s. No scores were recored."
+            % checkpoints_dir_path
+        )
+
+    ckpt_performance: list[tuple[float, str]] = sorted(results, reverse=True)
+
+    # (score, path)
+    return ckpt_performance[0]
+
+
+def _run_phased_training(
+    ctx,
+    train_args: TrainingArgs,
+    torch_args: TorchrunArgs,
+    base_dir: str,
+    phase1_data: str,
+    phase1_num_epochs: typing.Optional[int],
```

Comment:
nit: python now allows to say `int | None` instead.

Reply:
++ will do

Reply:
updated